sql/sql_db.cc

```diff
@@ -253,6 +253,37 @@ enum_db_read_only get_db_read_only(const dd::Schema &schema) {
   DBUG_RETURN(static_cast<enum_db_read_only>(val));
 }
 
+struct Update_thd_db_metadata : public Do_THD_Impl {
+  const char *db_name;
+  std::string metadata;
+
+  Update_thd_db_metadata(const char *db_name, const char *metadata)
+      : db_name(db_name), metadata(metadata) {}
+
+  virtual void operator()(THD *thd) {
+    const auto name_len = strlen(db_name);
+    if (name_len == thd->db().length &&
+        strncmp(thd->db().str, db_name, name_len) == 0) {
+      mysql_mutex_lock(&thd->LOCK_thd_db_metadata);
```

Comment:
I think you would need to acquire the LOCK_thd_data before checking thd->db().str? Or is thd->LOCK_thd_data acquired somehow to prevent db from changing?